dds/DCPS/DataSampleHeader.cpp

```diff
@@ -458,6 +457,149 @@ DataSampleHeader::join(const DataSampleHeader& first,
   return true;
 }
 
+OPENDDS_STRING to_string(const MessageId value)
+{
+  switch (value) {
+  case SAMPLE_DATA:
+    return "SAMPLE_DATA";
+  case DATAWRITER_LIVELINESS:
+    return "DATAWRITER_LIVELINESS";
+  case INSTANCE_REGISTRATION:
+    return "INSTANCE_REGISTRATION";
+  case UNREGISTER_INSTANCE:
+    return "UNREGISTER_INSTANCE";
+  case DISPOSE_INSTANCE:
+    return "DISPOSE_INSTANCE";
+  case GRACEFUL_DISCONNECT:
+    return "GRACEFUL_DISCONNECT";
+  case REQUEST_ACK:
+    return "REQUEST_ACK";
+  case SAMPLE_ACK:
+    return "SAMPLE_ACK";
+  case END_COHERENT_CHANGES:
+    return "END_COHERENT_CHANGES";
+  case TRANSPORT_CONTROL:
+    return "TRANSPORT_CONTROL";
+  case DISPOSE_UNREGISTER_INSTANCE:
+    return "DISPOSE_UNREGISTER_INSTANCE";
+  case END_HISTORIC_SAMPLES:
+    return "END_HISTORIC_SAMPLES";
+  default:
+    return "Unknown";
+  }
+}
+
+OPENDDS_STRING to_string(const SubMessageId value)
+{
+  switch (value) {
+  case SUBMESSAGE_NONE:
+    return "SUBMESSAGE_NONE";
+  case MULTICAST_SYN:
+    return "MULTICAST_SYN";
+  case MULTICAST_SYNACK:
+    return "MULTICAST_SYNACK";
+  case MULTICAST_NAK:
+    return "MULTICAST_NAK";
+  case MULTICAST_NAKACK:
+    return "MULTICAST_NAKACK";
+  default:
+    return "Unknown";
+  }
+}
+
+OPENDDS_STRING to_string(const DataSampleHeader& value)
+{
+  OPENDDS_STRING ret;
+  if (value.submessage_id_ != SUBMESSAGE_NONE) {
+    ret += to_string(SubMessageId(value.submessage_id_));
+    int sz = std::snprintf(NULL, 0, "%02x", unsigned(value.submessage_id_));
```

Comment:
Don't need a call to snprintf for "sizing" because the size is already determined by the format spec.